From 238dab224683098cafc3b41e9f63113ff14b810c Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Fri, 23 Jun 2017 09:30:29 -0400
Subject: [PATCH] Memory exhaustion in ReadCINImage

When identify CIN file that contains User defined data, imagemagick will allocate memory to store the
data in function ReadCINImage in coders\inc.c

There is a security checking in the function SetImageExtent,
but it after memory allocation, so IM can not control the memory usage

bug: https://github.com/ImageMagick/ImageMagick/issues/519
bug-debian: https://bugs.debian.org/867810
origin: https://github.com/ImageMagick/ImageMagick/commit/8e576918eb28501626040d925d19a0910f3dfac4

(cherry picked from commit 8e576918eb28501626040d925d19a0910f3dfac4)
---
 coders/cin.c | 3 +++
 coders/rle.c | 2 +-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/coders/cin.c b/coders/cin.c
index f394d8c00..1ef0ff296 100644
--- a/coders/cin.c
+++ b/coders/cin.c
@@ -710,6 +710,8 @@ static Image *ReadCINImage(const ImageInfo *image_info,ExceptionInfo *exception)
       /*
         User defined data.
       */
+      if (cin.file.user_length > GetBlobSize(image))
+        ThrowReaderException(CorruptImageError,"ImproperImageHeader");
       profile=BlobToStringInfo((const void *) NULL,cin.file.user_length);
       if (profile == (StringInfo *) NULL)
         ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
@@ -822,6 +824,7 @@ ModuleExport size_t RegisterCINImage(void)
   entry->encoder=(EncodeImageHandler *) WriteCINImage;
   entry->magick=(IsImageFormatHandler *) IsCIN;
   entry->adjoin=MagickFalse;
+  entry->seekable_stream=MagickTrue;
   entry->description=ConstantString("Cineon Image File");
   entry->module=ConstantString("CIN");
   (void) RegisterMagickInfo(entry);
diff --git a/coders/rle.c b/coders/rle.c
index 9945a88fa..b3f55090a 100644
--- a/coders/rle.c
+++ b/coders/rle.c
@@ -744,7 +744,7 @@ ModuleExport size_t RegisterRLEImage(void)
   entry=SetMagickInfo("RLE");
   entry->decoder=(DecodeImageHandler *) ReadRLEImage;
   entry->magick=(IsImageFormatHandler *) IsRLE;
-  entry->blob_support=MagickFalse;
+  entry->seekable_stream=MagickTrue;
   entry->adjoin=MagickFalse;
   entry->description=ConstantString("Utah Run length encoded image");
   entry->module=ConstantString("RLE");
